[Merged by Bors] - feat(Topology): `π₁(E⧸G) ≃* G` for `E` simply connected

alreadydone opened the pull request:
---

Further work originally highlighted by the author:
+ characterization of the homomorphism
+ change of basepoint `e : E` leads to conjugation by `G`
+ if `G` is abelian then independent of choice of basepoint (and we can remove the ᵐᵒᵖ)
+ $\pi_1(S^1)\cong\mathbb{Z}$
+ define winding number
+ generalize to `π₁(E⧸G)⧸π₁(E) ≃* G` for E path connected


- [x] depends on: #7596

<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

alreadydone committed:
feat(Topology): covering maps from properly discontinuous actions and discrete subgroups

alreadydone committed:
docstrings

alreadydone committed:
revise docstring

alreadydone committed:
minigolf

alreadydone committed:
refactor & to-do & docstring

alreadydone committed:
circle covers itself: isCoveringMap_nsmul

alreadydone committed:
fix bad search&replace

alreadydone committed:
Hmm how this #align got removed ...

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
fix merge

digama0 committed:
chore: bump lean toolchain

PatrickMassot committed:
Merge branch 'master' into CoveringMap_discrete

PatrickMassot committed:
Update lean-toolchain

alreadydone committed:
fix after merge

alreadydone committed:
Merge remote-tracking branch 'origin/master' into CoveringMap_discrete

alreadydone committed:
fix conflict

alreadydone committed:
IsLocallyHomeomorph -> IsLocalHomeomorph

alreadydone committed:
add import

alreadydone committed:
fix cyclic import

alreadydone committed:
Discard changes to Mathlib/Topology/Algebra/Group/Basic.lean

alreadydone committed:
fix

alreadydone committed:
fix AddCircle

alreadydone committed:
split file and fix all

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
fixes & edit module docstrings

alreadydone committed:
generalize `isCoveringMap_nsmul` and add `isCoveringMap_zsmul`

alreadydone committed:
fix import

alreadydone committed:
unused linter

alreadydone committed:
leftover

alreadydone committed:
shake

alreadydone committed:
minor

alreadydone committed:
Discard changes to Mathlib/Topology/Instances/AddCircle.lean

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
reapply changes to Topology/Instances/AddCircle/Defs + fixes

alreadydone committed:
Merge remote-tracking branch 'upstream/master' into CoveringMap_discrete

alreadydone committed:
rename

alreadydone committed:
fixes

alreadydone committed:
module + public import + expose

alreadydone committed:
address review

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
use Set.Nonempty in ProperlyDiscontinuousSMul

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
fix

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
strengthen to `IsQuotientCoveringMap`

alreadydone committed:
try remove manual to_additive translation

alreadydone committed:
fix

alreadydone committed:
rename group lemma

alreadydone committed:
rename ProperlyDiscontinuousSMul.disjoint_image_nhds

alreadydone committed:
address review

alreadydone committed:
Merge remote-tracking branch 'upstream/master' into CoveringMap_discrete

alreadydone committed:
`IsCoveringMap Circle.exp`

alreadydone committed:
trivial golf

alreadydone committed:
constructed hom from π₁(base) to group acting upstairs

alreadydone committed:
address review

alreadydone committed:
generalize isAddQuotientCoveringMap_zsmul/nsmul

alreadydone committed:
fix

alreadydone committed:
Merge branch 'master' into CoveringMap_discrete

alreadydone committed:
.Constructions

alreadydone committed:
WIP

alreadydone committed:
Merge branch 'master' into pi1_S1

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash committed:
Add proof of `fundamentalGroupToMulOpposite_surjective`

ocfnash committed:
Minor fixes

ocfnash committed:
Style tweak

ocfnash committed:
Add `IsQuotientCoveringMap.fundamentalGroupEquiv`

alreadydone commented on Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean:
```diff
@@ -65,6 +66,9 @@ instance (x y : X) : Subsingleton (Path.Homotopic.Quotient x y) :=
     rw [simply_connected_iff_unique_homotopic] at *; tauto))
 
+instance (x : X) : Subsingleton (FundamentalGroup X x) :=
+  show Subsingleton (Path.Homotopic.Quotient x x) from inferInstance
```
```suggestion
  inferInstanceAs <| Subsingleton (Path.Homotopic.Quotient x x)
```

ocfnash commented on Mathlib/Algebra/Group/Subgroup/Ker.lean:
```diff
@@ -67,6 +67,10 @@ open Subgroup
 def range (f : G →* N) : Subgroup N :=
   Subgroup.copy ((⊤ : Subgroup G).map f) (Set.range f) (by simp)
 
+@[to_additive]
```
splice merge

ocfnash replied:
splice-bot

ocfnash commented on Mathlib/AlgebraicTopology/FundamentalGroupoid/Basic.lean:
```diff
@@ -358,11 +358,11 @@ theorem fromPath_eq_iff_homotopic {x₀ x₁ : X} (f : Path x₀ x₁) (g : Path
   ⟨fun ih ↦ Quotient.exact ih, fun h ↦ Quotient.sound h⟩
 
 lemma eqToHom_eq {x₀ x₁ : X} (h : x₀ = x₁) :
-    eqToHom (congr_arg mk h) = ⟦(Path.refl x₁).cast h rfl⟧ := by subst h; rfl
```
splice-bot

ocfnash commented on Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean:
```diff
@@ -65,6 +66,9 @@ instance (x y : X) : Subsingleton (Path.Homotopic.Quotient x y) :=
   @Unique.instSubsingleton _ (Nonempty.some (by
     rw [simply_connected_iff_unique_homotopic] at *; tauto))
 
+instance (x : X) : Subsingleton (FundamentalGroup X x) :=
```
splice-bot

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash commented on Mathlib/Topology/Homotopy/Path.lean:
```diff
@@ -379,6 +379,10 @@ theorem cast_cast {x y : X} (γ : Homotopic.Quotient x y) {x' y'} (hx : x' = x)
   induction γ using Quotient.ind with | mk γ =>
   rfl
 
+theorem cast_heq {x y x' y' : X} (hx : x' = x) (hy : y' = y) {γ : Homotopic.Quotient x y} :
```
splice-bot

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash committed:
Review file `Topology.Covering.Quotient`.

Pushing review suggestions as commit

ocfnash committed:
Tweak `fundamentalGroupEquiv`

ocfnash commented on Mathlib/Topology/Covering/Quotient.lean:
```diff
@@ -46,27 +46,90 @@ namespace IsQuotientCoveringMap
 @[to_additive] theorem subgroup_congr (S S' : Subgroup G) (eq : S = S') :
     IsQuotientCoveringMap f S ↔ IsQuotientCoveringMap f S' := by rw [eq]
 
+variable {f G} (hf : IsQuotientCoveringMap f G)
```
Although I did technically push [a commit](https://github.com/leanprover-community/mathlib4/pull/33108/commits/b7fd02947cdfe591dc516321c3b222ced950ff09) which changed this file, I believe it is reasonable to regard those changes as equivalent to the sorts of changes which are often enacted via the GH suggestion feature. I am therefore comfortable to stretch a point and declare:

splice-bot merge

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash committed:
Fix `backward.isDefEq.respectTransparency false`

The fix is to add the `reducible` attribute to `_root_.toContinuousMap`.
Without this, at non-reducible transparency Lean is unable to see that
`Path.toContinuousMap` and `_root_.toContinuousMap` are defeq and problems
ensue.

An alternative fix is to write `γ.toContinuousMap` wherever necessary in
the definition `liftPathQuotient` but this is really avoiding the problem
rather than solving it.

The other changes to `liftPathQuotient` here are just intended to improve
readability.

alreadydone commented on Mathlib/Topology/Homotopy/Lifting.lean:
```diff
@@ -496,3 +535,139 @@ theorem IsCoveringMapOn.existsUnique_continuousMap_lifts [SimplyConnectedSpace A
+    FundamentalGroup X x ≃* G :=
+  (MulEquiv.ofBijective (hp.fundamentalGroupToMulOpposite e)
+    ⟨hp.fundamentalGroupToMulOpposite_injective e,
+     hp.fundamentalGroupToMulOpposite_surjective e⟩).trans <| (MulEquiv.inv' G).symm
```
If you want this to agree with #36119 I think you need to use [MulOpposite.opMulEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Equiv/Opposite.html#MulOpposite.opMulEquiv) (for G commutative) instead of `MulEquiv.inv'`.

ocfnash replied:
Thanks, yes. I'm am still considering what is best here and I agree it might not be this inversion.

ocfnash replied:
In the end I was persuaded to drop `MulEquiv.inv'`.

ocfnash commented on Mathlib/Topology/ContinuousMap/Defs.lean:
```diff
@@ -61,7 +61,7 @@ variable {F X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y] [FunLike F X
 
 /-- Coerce a bundled morphism with a `ContinuousMapClass` instance to a `ContinuousMap`. -/
-@[coe] def toContinuousMap (f : F) : C(X, Y) := ⟨f, map_continuous f⟩
+@[coe, reducible] def toContinuousMap (f : F) : C(X, Y) := ⟨f, map_continuous f⟩
```
I have proposed this in #40477

ocfnash committed:
Update Mathlib/AlgebraicTopology/FundamentalGroupoid/SimplyConnected.lean

Co-authored-by: Junyan Xu <junyanxu.math@gmail.com>

ocfnash committed:
Merge branch 'master' into pi1_S1

ocfnash committed:
Drop remaining `backward.isDefEq.respectTransparency`

We may need to think about redefining `Path.map` to produce
something of type `Path.Homotopic.Quotient y₀ y₁` subject to
the user providing proof that `f x₀ = y₀` and f x₁ = y₁`.

I think this would make proofs like this easier.

ocfnash committed:
Style tweaks

ocfnash committed:
Fix after adding `open Function`

ocfnash commented on Mathlib/AlgebraicTopology/FundamentalGroupoid/FundamentalGroup.lean:
```diff
@@ -32,17 +32,19 @@ variable (X)
 /-- The fundamental group is the automorphism group (vertex group) of the basepoint
 in the fundamental groupoid. -/
-def FundamentalGroup (x : X) :=
+abbrev FundamentalGroup (x : X) :=
   End (FundamentalGroupoid.mk x)
 
-instance (x : X) : Group (FundamentalGroup X x) := inferInstanceAs (Group (End _))
-
-instance (x : X) : Inhabited (FundamentalGroup X x) := inferInstanceAs (Inhabited (End _))
-
```
I think the current situation where `FundamentalGroup` is an `expose`d `def` is the worst of all worlds and that we should either implement this change (i.e., make it an `abbrev`) or else we should make it an unexposed `def`, add:
```lean
-- This seems to be missing so would presumably be added upstream
instance (x : X) : Group (Path.Homotopic.Quotient x x) :=
  inferInstanceAs <| Group <| End <| FundamentalGroupoid.mk x

def equivHomotopicPath (x : X) : FundamentalGroup X x ≃* Path.Homotopic.Quotient x x :=
  MulEquiv.refl _
```
and then think carefully downstream about whether each lemma should make a statement about `γ : Path.Homotopic.Quotient x x` or `γ : FundamentalGroup X x`.

I briefly experimented with this latter approach and was left unconvinced it was a change for the better but I am happy to follow others' opinions.

ocfnash commented on Mathlib/Topology/Homotopy/Lifting.lean:
```diff
@@ -354,44 +356,90 @@ theorem homotopicRel_iff_comp [PreconnectedSpace A] {f₀ f₁ : C(A, E)} {S : S
+  γ.hrecOn g hg
+
+theorem map_liftPathQuotient {x y : X} (γ : Path.Homotopic.Quotient x y) (e : p ⁻¹' {x}) :
+    (cov.liftPathQuotient γ e).map ⟨p, cov.continuous⟩ = γ.cast e.2 (cov.monodromy γ e).2 := by
```
I'll note in passing that I think we should probably restate [Path.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Path.html#Path.map) and so that instead of consuming `x₀`, `x₁` and producing a term of type `Path (f x₀) (f x₁)` it consumes `x₀`, `x₁`, `y₀`, `y₁` together with proofs that `y₀ = f x₀`, `y₁ = f x₁` and produce a term of type `Path y₀ y₁`.

Likewise for [Path.Homotopic.Quotient.map](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Homotopy/Path.html#Path.Homotopic.Quotient.map) (used here).

jcommelin reviewed (commented):
Thanks :tada:

bors merge
